### PR TITLE
[SPARK-34376][SQL] Support regexp as a SQL function

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

We have equality in `SqlBase.g4` for `RLIKE: 'RLIKE' | 'REGEXP';`
We seemed to miss adding` REGEXP` as a SQL function just like` RLIKE`

### Why are the changes needed?

symmetry and beauty
This is also a builtin  function in Hive, we can reduce the migration pain for those users

### Does this PR introduce _any_ user-facing change?

yes new regexp function as an alias as rlike

### How was this patch tested?

new tests

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu thanks for checking

File: sql/core/src/test/resources/sql-tests/inputs/regexp-functions.sql [hunk 1]
```diff
@@ -46,4 +46,9 @@ SELECT regexp_replace('healthy, wealthy, and wise', '\\w', 'something', null);
 
 -- regexp_like
 SELECT regexp_like('1a 2b 14m', '\\d+b');
-SELECT regexp_like('1a 2b 14m', '[a-z]+b');
\ No newline at end of file
+SELECT regexp_like('1a 2b 14m', '[a-z]+b');
+
```
Comment (jaceklaskowski) [inline sql/core/src/test/resources/sql-tests/inputs/regexp-functions.sql:50]: nit: remote the empty line

Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/inputs/regexp-functions.sql:50]: good point

**Comment by maropu** [comment]
Adding it looks okay. Any other DBMS supports the func?

**Comment by yaooqinn** [comment]
> Adding it looks okay.  Any other DBMS supports the func?

I found many others support only the form `subject RLIKE/REGEX pattern` and with a synonym `REGEXP_LIKE(subject, pattern)` used as a function.
https://dev.mysql.com/doc/refman/8.0/en/regexp.html#operator_regexp

https://docs.snowflake.com/en/sql-reference/functions-regexp.html snowflake also has RLIKE(a, b, c) 😂

https://www.postgresql.org/docs/9.0/functions-string.html in Postgres it is `regexp_matches(string text, pattern text [, flags text])` 😂

For the situation that we have in Spark now, I hope we can get this function in, as we already have `rlike` as SQL regular functions.

And in real-world cases, users may always deploy Spark and Hive together, this small change can remove a big barrier for miraging Hive to Spark SQL, which is very likely to happen all around the world these days. For example, I am going to migrating 20k hive jobs to Spark next month after the Chinese new year festival 😂

**Review by maropu — APPROVED** [review]
cc: @HyukjinKwon

**Comment by maropu** [comment]
retest this plase

**Comment by maropu** [comment]
retest this please

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**